frontend/catalyst/pennylane_extensions.py

```diff
@@ -2633,22 +2635,38 @@ def callback_implementation(
     flat_args, in_tree = tree_flatten((args, kwargs))
     metadata = CallbackClosure(args, kwargs)
 
+    def to_shaped_array(ty):
+        if ty == inspect.Signature.empty:
+            return None
+        if isinstance(ty, ShapedArray):
+            return ty
+        return shaped_abstractify(ty)
+
+    results_aval = tree_map(to_shaped_array, result_shape_dtypes)
+    flat_results_aval, out_tree = tree_flatten(results_aval)
+
     def _flat_callback(flat_args):
         """Each flat_arg is a pointer.
 
         It is a pointer to a memref object.
         To find out which element type it has, we use the signature obtained previously.
         """
         jnpargs = metadata.getArgsAsJAXArrays(flat_args)
-
         args, kwargs = tree_unflatten(in_tree, jnpargs)
-        return tree_leaves(cb(*args, **kwargs))
+        retvals = tree_leaves(cb(*args, **kwargs))
+        return_values = []
+        for retval in retvals:
+            ranked_memref = get_ranked_memref_descriptor(retval)
+            element_size = ctypes.sizeof(ranked_memref.aligned.contents)
+            unranked_memref = get_unranked_memref_descriptor(retval)
+            unranked_memref_ptr = ctypes.cast(ctypes.pointer(unranked_memref), ctypes.c_void_p)
+            # We need to keep a value of retval around
+            # Otherwise, Python's garbage collection will collect the memory
+            # before we run the memory copy in the runtime.
+            # We need to copy the unranked_memref_ptr and we need to know the element size.
+            return_values.append((unranked_memref_ptr, element_size, retval))
+        return return_values
```

Comment:
Don't you need to unflatten the returns?

Reply:
They are already unflattened in line 2673. The "unflattened" is only useful for the tracing. The generated program will still execute the flattened list. (I am attaching a screenshot because it is difficult to get a link). Happy to answer more questions here.

![Screenshot from 2024-04-01 15-29-03](https://github.com/PennyLaneAI/catalyst/assets/110487834/5091d9cc-19ac-4da3-a76d-5313115c8044)